### PR TITLE
Speech and Braille: don't report many clickables in a row in browse mode.

### LINKED ISSUE #7430: NVDA repeats the word "clickable" several times in a row before just saying what the item is.
In some pages, NVDA repeats the word clickable several times before saying what the clickable item is. Something like: "clickable clickable clickable clickable clickable clickable clickable clickable click here to begin."
For example: go to `my.snhu.edu`. From the top of the page, press `f` to go to the next form field.
NVDA uses the word "clickable" eight times in the course of saying I landed on an edit box that is titled: "search this site." This is, I am sure, because of bad web design, but NVDA should cope with this and not say the word "clickable" more than once per item, and never for form controls, at least, it seems obvious that a button or text field is clickable, this is how one interacts with them.

**Edit**
Since I am not able to reproduce this via `snhu.edu` site, consider the following example (paste into the address bar in Firefox or Chrome):
`data:text/html,<p>text before the clickable div</p><div onClick=";"><div onClick=";">This is some text before the button <br /><button>foo</button></div> bar</div><p>This is the paragraph</p>`

- After NVDA reads out "text before the clickable div"
- Press `f` to jump to the button.

### Actual behaviour:
Chrome
```
clickable
foo
button
```
Firefox
```
clickable
clickable
foo
button
```

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
Fixes #7430

### Summary of the issue:
NVDA reports certain fields in browse mode as 'clickable' if the author has provided an onclick handler and NVDA would not normally report anything else for that field (E.g. has no reportable role).
However, with the increased use of onclick handlers on web pages, there are many cases where NVDA will report multiple clickables in a row E.g. "clickable clickable clickable clickable". This is extremely annoying.
To make things worse, versions of Chrome earlier than 63 also propagated the existence of the click action to all descendant accessibles, meaning that NVDA would report clickable for every descendant of a clickable node. Although this was fixed in Chrome a while back, many 3rd party apps that use Chromium / Electron (such as Microsoft Teams) still have this issue as they are using an older version of Chromium.
 
### Description of how this pull request fixes the issue:
Several solutions were described in issue #7430. This PR goes with solution 3 (collapsing clickables).
Although it wasn't the most perfect solution, it was the easiest to implement, and probably the easiest to follow code-wise. It also hopefully provides a good balance between usability and predictability. It is certainly a major improvement over what we currently have, but also erring on the side of reporting more than less.
In short: in both speech and braille now, NVDA will never report multiple clickables in a row without any other intervening text. As NVDA enters one or more clickable fields, it will only report clickable once. However, if one or more of these fields are exited, and then 1 or more new clickable fields are entered, NVDA will again report clickable, but only once when entering those new fields. To put it another way, NVDA will only report clickable once for any run of field starts.

 ### Testing performed:
Tested example in #7430. Multiple clickables in a row are now only announced as one clickable.
Also tested 
[clickables.html.txt](https://github.com/nvaccess/nvda/files/2551843/clickables.html.txt)
which contains various different paragraphs and divs containing single and nested clickable spans.

### Known issues with pull request:
None known.

### Change log entry:
Bug fixes:
* NVDA no longer reports clickable multiple times in a row when navigating clickable content in Browse Mode. (#7430)

### CODE REVIEW

File: source/braille.py [hunk 1]
```diff
@@ -872,7 +870,21 @@ def _addTextWithFields(self, info, formatConfig, isSelection=False):
 					if self._skipFieldsNotAtStartOfNode and not field.get("_startOfNode"):
 						text = None
 					else:
+						textList=[]
+						states=field.get('states')
+						if states and controlTypes.STATE_CLICKABLE in states:
+							if not inClickable:
+								# We have entered an outer most clickable or entered a new clickable after exiting a previous one 
+								# Report it if there is nothing else interesting about the field, but not if the user turned it off. 
+								if formatConfig['reportClickable']:
```
Comment (leonardder) [inline source/braille.py]: Would it be possible to check this earlier on? I guess it is not necessary to check the clickable state if we don't report clickables at all.

File: source/speech.py [hunk 2]
```diff
@@ -853,9 +853,20 @@ def speakTextInfo(info,useCache=True,formatConfig=None,unit=None,reason=controlT
 				isTextBlank=False
 				_speakTextInfo_addMath(speechSequence,info,field)
 
+	# When true, we are inside a clickable field, and should therefore not announce any more new clickable fields
```
Comment (leonardder) [inline source/speech.py]: Could you also provide this comment in braille.py when declaring inClickable?

File: source/speech.py [hunk 3]
```diff
@@ -853,9 +853,20 @@ def speakTextInfo(info,useCache=True,formatConfig=None,unit=None,reason=controlT
...
+	inClickable=False
 	#Get speech text for any fields in the new controlFieldStack that are not in the old controlFieldStack
 	for count in xrange(commonFieldCount,len(newControlFieldStack)):
 		field=newControlFieldStack[count]
+		if not inClickable:
+			states=field.get('states')
+			if states and controlTypes.STATE_CLICKABLE in states:
+				# We entered the most outer clickable, so announce it, if we won't be announcing anything else interesting for this field, but not if the user turned them off.
+				if formatConfig['reportClickable']:
```
Comment (leonardder) [inline source/speech.py]: Same question as for de braille module.

File: source/speech.py [hunk 4]
```diff
@@ -915,13 +928,30 @@ def speakTextInfo(info,useCache=True,formatConfig=None,unit=None,reason=controlT
...
 				# Control fields always start a new chunk, even if they have no field text.
 				inTextChunk=False
-				fieldText=info.getControlFieldSpeech(command.field,newControlFieldStack,"start_relative",formatConfig,extraDetail,reason=reason)
+				tempTextList=[]
+				states=command.field.get('states')
+				if states and controlTypes.STATE_CLICKABLE in states:
+					if not inClickable:
+						# We have entered an outer most clickable or entered a new clickable after exiting a previous one 
+						# Announce it if there is nothing else interesting about the field, but not if the user turned it off. 
+						if formatConfig['reportClickable']:
```
Comment (leonardder) [inline source/speech.py]: Same concern as above.

**Comment by michaelDCurran** [comment]
@leonardder: I have addressed your review comments, and also somewhat 
simplified the logic, which actually fixes a particular corner case 
which was breaking.

Now we simply report the first clickable in a run of control starts, and 
text or a control end breaks the run.

Now all tests in clickables.html in both speech and braille work as 
expected. Previously the "Clickable spans in middle of clickable 
paragraph" test was not correct in braille.

File: source/speech.py [hunk 5]
```diff
@@ -915,9 +930,24 @@ def speakTextInfo(info,useCache=True,formatConfig=None,unit=None,reason=controlT
...
-				fieldText=info.getControlFieldSpeech(command.field,newControlFieldStack,"start_relative",formatConfig,extraDetail,reason=reason)
+				tempTextList=[]
+				if not inClickable and formatConfig['reportClickable']:
+					states=command.field.get('states')
+					if states and controlTypes.STATE_CLICKABLE in states:
+						# We have entered an outer most clickable or entered a new clickable after exiting a previous one 
+						# Announce it if there is nothing else interesting about the field, but not if the user turned it off. 
+						presCat=command.field.getPresentationCategory(newControlFieldStack[0:],formatConfig,reason)
+						if not presCat or presCat is command.field.PRESCAT_LAYOUT:
+							tempTextList.append(controlTypes.stateLabels[controlTypes.STATE_CLICKABLE])
```
Comment (derekriemer) [inline source/speech.py]: What if we keep a list of roles where all of that type of role is inherently clickable. That way we don't say something is a clibkable link or a clickable button? Or does this code already do that?

File: source/speech.py [hunk 6]
```diff
@@ -915,9 +930,24 @@ def speakTextInfo(info,useCache=True,formatConfig=None,unit=None,reason=controlT
...
+						inClickable=True
+				text=info.getControlFieldSpeech(command.field,newControlFieldStack,"start_relative",formatConfig,extraDetail,reason=reason)
+				if text:
+					tempTextList.append(text)
```
Comment (derekriemer) [inline source/speech.py]: Should text being true turn a run of clickables off, or do I misunderstand this? I'm reading it pretty fast.

File: source/speech.py (hunk 5 above)
Comment (michaelDCurran) [reply source/speech.py]: We are already doing this by checking the field's presentationCategory.

File: source/speech.py (hunk 6 above)
Comment (michaelDCurran) [reply source/speech.py]: Nope. That text variable just holds  the speech for that particular control Start.

**Review by LeonarddeR — APPROVED** [review]
Looks way cleaner now, thanks!

**Pull request merged**